### PR TITLE
Ge v1

### PR DESCRIPTION
**Pull request by jjsjann123**
This is a second attempt to use graph executor to run forward on a gradient. This allows a secondary chance to profile intermediate tensor introduced by autodiff.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This is an alternative implementation to #47434, where plumbing the optimized graph through interpreter turns out to be hitting some messy issues in both tests and infrustructure.

**Comment by jjsjann123** [comment]
Muhahahahaha, this looks really promising.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -458,7 +468,6 @@ void ProfilingGraphExecutorImpl::runProfilingOptimizations(
       // graph is essentially in a torch.no_grad context.
       UpdateDifferentiableGraphRequiresGrad(gradient.f, false);
       GRAPH_DEBUG("After UpdateDifferentiableGraphRequiresGrad ", *gradient.f);
-      runDiffGraphPasses(gradient.f);
```
Comment (Krovatkin) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:461]: Oh man, I'm missed this two times I was reading this code!

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 2]
```diff
@@ -501,6 +510,13 @@ GraphExecutor* getGradExecutor(Operation& op) {
   }
   return nullptr;
 }
+
+GraphExecutor* getDifferentiableGraphOpExecutor(Operation& op) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/runtime/graph_executor.cpp:514]: Good idea!

File: torch/jit/_fuser.py [hunk 3]
```diff
@@ -64,7 +65,37 @@ def fuser(name):
...
+
+        assert(len(fw_states) == len(diff_nodes))
+        # swap each differentiable graph with optimized graph in their execution plan
+        for n, state in zip(diff_nodes, fw_states):
+            fw_execution_plans = list(state.execution_plans.values())
+            assert(len(fw_execution_plans) == 1)
+            n.g_('Subgraph', fw_execution_plans[0].graph)
+
+        return graph
+    except Exception:
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/jit/_fuser.py:97]: This is intended.

The code above asserts on: 1. a unique execution plan for each differentiable ops; 2. `getDebugState()` only works when we have optimized_plan.
Under those scenarios, we would still want to return the vanilla graph instead of assert and crash.

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 4]
```diff
@@ -368,13 +368,14 @@ struct DifferentiableGraphBackward : public autograd::Node {
...
       : f(grad.f, "<foward op>"),
+        legacy_f(grad.f, "<foward op>"),
         grad(std::move(grad)),
         grad_executor(this->grad.df, "<backward op>"),
         num_inputs(this->grad.f->inputs().size()),
         num_outputs(this->grad.f->outputs().size()) {}
 
   // XXX: keep in mind that stack can be larger than the inputs we need!
-  void operator()(Stack* stack) const {
+  void operator()(Stack* stack) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/runtime/graph_executor.cpp:378]: Let me try to put it back.

In the other PR, I was updating member variable inside this function and had to remove the const flag. I don't think we need it any more.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:461]: I missed the corresponding part in legacy executor and it took me a while to root cause the segfault in CI previously. But there's really no good solution there for legacy executor, hence the branching in `DifferentiableGraphOp::operator()`. 😢

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 5]
```diff
@@ -392,7 +393,13 @@ struct DifferentiableGraphOp {
     }
 
     detachVariables(*stack);
-    InterpreterState(f).run(*stack);
+    if (IsNewExecutorEnabled()) {
+      ExecutionPlan plan =
+          f_ptr->getPlanFor(*stack, GraphExecutor::getDefaultNumBailOuts());
+      InterpreterState(plan.code).run(*stack);
+    } else {
+      InterpreterState(legacy_f).run(*stack);
```
Comment (eellison) [inline torch/csrc/jit/runtime/graph_executor.cpp:401]: unrelated to this PR: is it not expensive that we're reinstantiating `InterpreterState` on every forward ?

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 6]
```diff
@@ -33,11 +33,13 @@
 #include <torch/csrc/jit/passes/update_differentiable_graph_requires_grad.h>
 #include <torch/csrc/jit/passes/utils/subgraph_utils.h>
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
 C10_DEFINE_bool(
     torch_jit_enable_new_executor,
     true,
     "If this flag is set to false TorchScript will be using the legacy/original executor");
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:42]: thanks for doing this !!

**Review by eellison — COMMENTED** [review]
Looks good, I want to know more about the inserted `RemoveProfileNodesAndSpecializeTypes` call before landing

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 7]
```diff
@@ -452,13 +462,13 @@ void ProfilingGraphExecutorImpl::runProfilingOptimizations(
       GRAPH_DEBUG("After guardDifferentiableGraph:\n", *copy);
       auto diff_graph = std::move(dnode->g(attr::Subgraph));
       Gradient gradient = differentiate(diff_graph);
+      RemoveProfileNodesAndSpecializeTypes(gradient.f);
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:465]: I wasn't paying enough attention here, yeah the intention is just to remove profiling nodes. Will update.

File: torch/csrc/jit/runtime/graph_executor.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/graph_executor.cpp:401]: I'll have to forward this to @Krovatkin

File: test/test_jit_fuser_te.py [hunk 8]
```diff
@@ -950,8 +951,8 @@ def doit(x, y):
     def test_lstm(self):
         for device in self.devices:
             inputs = get_lstm_inputs(device, training=True)
-            module = self.checkScript(LSTMCellS, inputs)
-            self.assertLastGraphAllFused()
+            module = self.checkScript(LSTMCellS, inputs, extra_profile_runs=True)
+            self.assertAllFused(module.graph_for(inputs))
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply test/test_jit_fuser_te.py:955]: Because we can no longer rely on last_executed_optimized_graph to query optimized graph within a DifferentiableGraph.
Use `graph_for` allows us to query the updated subgraph and display it properly: https://github.com/pytorch/pytorch/pull/52136/files#diff-6f4bcfe1c1247ea7f4287e02d4bbc0381ba54828ee99bdd8a7c207fd79441470R94

File: torch/csrc/jit/runtime/graph_executor.cpp (hunk 5 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/graph_executor.cpp:401]: I don't know. I'll measure it when I get a chance.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp]: it probably worked because Profiler resets type information, but @eellison is right, we should just be using `removeProfilingNodes` or friends.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 9]
```diff
@@ -58,15 +62,41 @@ C10_DEFINE_int64(
 namespace torch {
 namespace jit {
 
+namespace {
+void removeProfileNodes(Block* b) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:66]: This exists [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/insert_guards.cpp#L8) (maybe there's a better location for it.)

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 10]
```diff
@@ -452,13 +482,13 @@ void ProfilingGraphExecutorImpl::runProfilingOptimizations(
       GRAPH_DEBUG("After guardDifferentiableGraph:\n", *copy);
       auto diff_graph = std::move(dnode->g(attr::Subgraph));
       Gradient gradient = differentiate(diff_graph);
+      removeProfileNodes(gradient.f);
       GRAPH_DEBUG("Forward graph:\n", *(gradient.f));
```
Comment (eellison) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:486]: Can we also be sure that we're removing specialized types here ? `removeTensorTypeSpecializations`

**Review by eellison — COMMENTED** [review]
Looks good, we should make sure we're not leaving in type specializations..

**Comment by jjsjann123** [comment]
I'll address @eellison 's comments, meanwhile, @Krovatkin anything relevant in the internal CI failures?

**Review by eellison — APPROVED** [review]
Thanks!
